scylla/src/transport/connection.rs

```diff
@@ -274,14 +283,26 @@ impl NonErrorQueryResponse {
             }
         };
 
-        Ok(QueryResult {
-            rows,
-            warnings: self.warnings,
-            tracing_id: self.tracing_id,
+        Ok((
+            QueryResult {
+                rows,
+                warnings: self.warnings,
+                tracing_id: self.tracing_id,
+                col_specs,
+                serialized_size,
+            },
             paging_state,
-            col_specs,
-            serialized_size,
-        })
+        ))
+    }
+
+    pub(crate) fn into_query_result(self) -> Result<QueryResult, QueryError> {
+        let (result, paging_state) = self.into_query_result_and_paging_state()?;
+
+        if !paging_state.finished() {
+            warn!("Internal driver API misuse: nonfinished paging state discarded by `NonErrorQueryResponse::into_query_result`");
+        }
```

Comment:
As far as I can tell this is used only for batches (and in one test, but it doesn't matter here), and batches don't have paging. In that case, if we encounter this warning this mean that either
- There is some serious bug in the driver
- There is bug in Scylla

both those cases warrant a panic imo.